state_server.go

```diff
@@ -0,0 +1,185 @@
+			LoopSubServer:    newSubServerStatus(),
+		},
+	}
+
+	options := []grpcweb.Option{
+		grpcweb.WithWebsockets(true),
+		grpcweb.WithWebsocketPingInterval(2 * time.Minute),
+		grpcweb.WithCorsForRegisteredEndpointsOnly(false),
+	}
+	s.grpcServer = grpc.NewServer()
+	s.grpcWebProxy = grpcweb.WrapServer(s.grpcServer, options...)
+
+	litrpc.RegisterStatusServer(s.grpcServer, s)
+
+	return s
+}
+
+// Stop stops the status server.
+func (s *statusServer) Stop() error {
+	s.grpcServer.Stop()
+	return nil
+}
+
+// GetSubServerState queries the current status of a given sub-server.
+//
+// NOTE: this is part of the litrpc.StatusServer interface.
+func (s *statusServer) GetSubServerState(_ context.Context,
+	_ *litrpc.GetSubServerStatusReq) (*litrpc.GetSubServerStatusResp,
+	error) {
+
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	resp := make(map[string]*litrpc.SubServerStatus, len(s.m))
+	for server, status := range s.m {
+		resp[server] = &litrpc.SubServerStatus{
+			Running: status.running,
+			Error:   status.err,
+		}
+	}
+
+	return &litrpc.GetSubServerStatusResp{
+		SubServers: resp,
+	}, nil
+}
+
+// getSubServerState queries the current status of a given sub-server.
+func (s *statusServer) getSubServerState(name string) (bool, string) {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+
+	system, ok := s.m[name]
+	if !ok {
+		return false, ""
+	}
+
+	return system.running, system.err
+}
+
+// setServerRunning can be used to set the status of a sub-server as running
+// with no errors.
+func (s *statusServer) setServerRunning(name string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.m[name] = &subServerStatus{
+		running: true,
+	}
+}
+
+// setServerStopped can be used to set the status of a sub-server as not running
+// and with no errors.
+func (s *statusServer) setServerStopped(name string) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	s.m[name] = &subServerStatus{
+		running: false,
+	}
+}
+
+// setServerErrored can be used to set the status of a sub-server as not running
+// and also to set an error message for the sub-server.
+func (s *statusServer) setServerErrored(name string, errStr string,
+	params ...interface{}) {
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	err := fmt.Sprintf(errStr, params...)
+	log.Errorf("could not start the %s sub-server: %s", name, err)
+
+	s.m[name] = &subServerStatus{
+		running: false,
+		err:     err,
+	}
+}
+
+// isHandling checks if the specified request is something to be handled by the
+// status server. If true is returned, the call was handled by the status
+// server and the caller MUST NOT handle it again. If false is returned, the
+// request was not handled and the caller MUST handle it.
+func (s *statusServer) isHandling(resp http.ResponseWriter,
+	req *http.Request) bool {
+
+	if !strings.Contains(req.URL.Path, "/litrpc.Status") {
+		return false
+	}
+
+	// gRPC web requests are easy to identify. Send them to the gRPC
+	// web proxy.
+	if s.grpcWebProxy.IsGrpcWebRequest(req) ||
```

Comment:
Again, it feels like this should just be hooked up in the right order to allow the existing `rpcProxy` to do this.